### ISSUE TITLE
Naive and slow implementation

### ISSUE THREAD

**Issue by TestPolygon** [sindresorhus/set-immediate-shim#4]
The description says that it is "Simple `setImmediate` module".
And people use it (over 2kk installations). They think that it's simple _and good_ one, but in fact it's simple _and slow_ one. But there is no any note about it. A lot of people do not check the code of modules that they install, they expect that the creator spent time to investigate the problem to create the best decision, but it is not about this repo. The word "simple" is not the correct description, but "simple and naive", or "simple and slow" are.

Using this code in the browser leads to a performance hit. **It's a fact.** For some cases the impact is significant.
For example: https://github.com/Stuk/jszip/issues/617

Here is the example of using of _the properly created_ `setImmediate`:
https://jphpsf.github.io/setImmediate-shim-demo/
**296ms** (`setImmediate`) vs **12617ms** (`setTimeout` **=== your implementation**)

Node.js has `setImmediate`. Using your implementation in the browser is the same thing as using `setTimeout`, but with the misleading name.

**Comment by sindresorhus** [comment]
You're right. My use-case for making this was not in hot code, so the `setInterval` overhead didn't matter. When I made this, the shims were all bloated and buggy. What is currently the best shim? I'm happy to recommend that instead.

**Comment by TestPolygon** [comment]
Here is the examples of `setImmediate` implementation: 
[1] https://github.com/YuzuJS/setImmediate/blob/master/setImmediate.js
and from core-js:
[2] https://github.com/zloirock/core-js/blob/master/packages/core-js/internals/task.js

I think it will be enough if you just add the implementation via `MessageChannel` ([it will work in WebWorkers too (in comparison with `postMessage`)](https://github.com/zloirock/core-js/blob/master/packages/core-js/internals/task.js#L70) for the browsers (with `setTimeous` as a fallback).
It has the good browser support: https://caniuse.com/#search=MessageChannel

In this case your implementation will be really simple (in comparison with others) and works well in most of cases.

Also: https://github.com/Stuk/jszip/pull/619#issuecomment-530724686

**Comment by sindresorhus** [comment]
> I think it would be enough if you just add the implementation via MessageChannel (it will work in WebWorkers too (in comparison with postMessage) for the browsers (with setTimeous as a fallback).

I agree. That's a good plan. I don't have time to work on this right now. Contributions are of course welcome. In the meantime, I added a note to the readme.

**Comment by AlttiRi** [comment]
Still not fixed?
Okay...
```js
export const setImmediate = /*#__PURE__*/ (function() {
    const {port1, port2} = new MessageChannel();
    const queue = [];

    port1.onmessage = function() {
        const callback = queue.shift();
        callback();
    };

    return function(callback) {
        port2.postMessage(null);
        queue.push(callback);
    };
})();
```
This implementation is only for browsers, to use it in Node.js it requires to call `ref`/`unref` for the proper process exit, but Node.js already has `setImmediate`, so in Node use the native `setImmediate`.

**Comment by AlttiRi** [comment]
I don't know why @Stuk still not changed `setImmediate` library to an other one while a lot of people ask him to do it. Okay, the other way to fix JSZip library is to fix _[this](https://github.com/sindresorhus/set-immediate-shim/)_ library if he do not want to use any other.

**Comment by AlttiRi** [comment]
I'm sorry, but why not?
```js
"use strict";
module.exports = /*#__PURE__*/ (function() {
    require("setimmediate");
    return setImmediate;
}());
```
It's better to do a wrapper for the other properly working library, than misname `setTimeout` with `setImmediate` for millions of people.

People specially use this library to get `setImmediate` instead of just using of `setTimeout`, but they get `setTimeout` anyway. I think it's not OK, it's wrong.

In any case "setimmediate" library adds `setImmediate` function to the global scope, but that wrapper just is a module. So it makes sense to have it.

---

UPD.
Ops, it's not working (as a module).

It should be OK:
```js
"use strict";
module.exports = /*#__PURE__*/ (function() {
    return require("./setImmediate-fixed.js").setImmediate;
}());
```
But it requires to modify [the original file](https://github.com/YuzuJS/setImmediate/blob/master/setImmediate.js)

[This:](https://github.com/YuzuJS/setImmediate/blob/f1ccbfdf09cb93aadf77c4aa749ea554503b9234/setImmediate.js#L4-L6
)
```js
    if (global.setImmediate) {
        return;
    }
```
replace with 
```js
    if (global.setImmediate) {
        module.exports = {
            setImmediate: global.setImmediate,
            clearImmediate: global.clearImmediate
        };
        return;
    }
```
and [this:](https://github.com/YuzuJS/setImmediate/blob/f1ccbfdf09cb93aadf77c4aa749ea554503b9234/setImmediate.js#L184-L185)
```js
    attachTo.setImmediate = setImmediate;
    attachTo.clearImmediate = clearImmediate;
```
replace with
```js
    module.exports = {setImmediate, clearImmediate};
```
